tools/deadlock_detector_example.txt

```diff
@@ -0,0 +1,315 @@
+#include <thread>
+
+int main(void) {
+  std::mutex m1;
+  std::mutex m2;
+  std::mutex m3;
+  std::mutex m4;
+
+  std::cout << "&m1: " << (void*)&m1 << std::endl;
+  std::cout << "&m2: " << (void*)&m2 << std::endl;
+  std::cout << "&m3: " << (void*)&m3 << std::endl;
+  std::cout << "&m4: " << (void*)&m4 << std::endl;
+
+  std::cout << "pid: " << getpid() << std::endl;
+  std::cout << "sleeping for a bit to allow trace to attach..." << std::endl;
+  std::this_thread::sleep_for(std::chrono::seconds(10));
+  std::cout << "starting program..." << std::endl;
+
+  auto t1 = std::thread([&m1, &m2] {
+    std::lock_guard<std::mutex> g1(m1);
+    std::lock_guard<std::mutex> g2(m2);
+  });
+  t1.join();
+
+  auto t2 = std::thread([&m2, &m3] {
+    std::lock_guard<std::mutex> g2(m2);
+    std::lock_guard<std::mutex> g3(m3);
+  });
+  t2.join();
+
+  auto t3 = std::thread([&m3, &m4] {
+    std::lock_guard<std::mutex> g3(m3);
+    std::lock_guard<std::mutex> g4(m4);
+  });
+  t3.join();
+
+  auto t4 = std::thread([&m1, &m4] {
+    std::lock_guard<std::mutex> g4(m4);
+    std::lock_guard<std::mutex> g1(m1);
+  });
+  t4.join();
+
+  std::cout << "sleeping to allow trace to collect data..." << std::endl;
+  std::this_thread::sleep_for(std::chrono::seconds(5));
+  std::cout << "done!" << std::endl;
+}
+```
+
+Note that an actual deadlock did not occur, although this mutex lock ordering
+creates the possibility of a deadlock, and this is a hint to the programmer to
+reconsider the lock ordering.
+
+
+# ./deadlock_detector.py /path/to/program $(pidof program) --dump-graph graph.json
+
+Tracing... Hit Ctrl-C to end.
+Nodes: 0, Edges: 0, Looking for cycle took 0.000062 seconds
+Nodes: 0, Edges: 0, Looking for cycle took 0.000066 seconds
+Nodes: 0, Edges: 0, Looking for cycle took 0.000065 seconds
+Nodes: 0, Edges: 0, Looking for cycle took 0.000053 seconds
+Nodes: 102, Edges: 4936, Looking for cycle took 0.155751 seconds
+Nodes: 102, Edges: 4951, Looking for cycle took 0.141393 seconds
+Nodes: 102, Edges: 4951, Looking for cycle took 0.119585 seconds
+Nodes: 102, Edges: 4951, Looking for cycle took 0.118088 seconds
+^C
+
+If the program does not find a deadlock, it will keep running until you hit
+Ctrl-C. It will also dump statistics about the number of nodes and edges in
+the mutex wait graph. If you want to serialize the graph to analyze it later,
+you can pass the `--dump-graph FILE` flag, and the program will serialize
+the graph in json format.
+
+
+# ./deadlock_detector.py /path/to/program $(pidof program) --lock-symbols custom_mutex1_lock,custom_mutex2_lock --unlock_symbols custom_mutex1_unlock,custom_mutex2_unlock
+
+Tracing... Hit Ctrl-C to end.
+Nodes: 0, Edges: 0, Looking for cycle took 0.000062 seconds
+Nodes: 0, Edges: 0, Looking for cycle took 0.000066 seconds
+Nodes: 0, Edges: 0, Looking for cycle took 0.000065 seconds
+Nodes: 0, Edges: 0, Looking for cycle took 0.000053 seconds
+Nodes: 102, Edges: 4936, Looking for cycle took 0.155751 seconds
```

Comment:
- I'll get rid of the time, it doesn't really help much. 
- I think it will be better to print the final graph size (number of mutexes and edges) only if the user presses Ctrl-C, or the program finds a deadlock.